src/link.js

```diff
@@ -0,0 +1,259 @@
+	 *	             if it's correct.   |                   |
+	 *	                                |                   |
+	 *	                      +---------+--------+          |
+	 *	Panel.urlInput#value  | Value validation |          |  User clicked "Link" in
+	 *	       is validated.  +---------+--------+          |  the toolbar. Retrieving
+	 *	                                |                   |  URL from Document and setting
+	 *	             PanelModel fires   |                   |  PanelModel#url.
+	 *	          PanelModel#execute.   +                   v
+	 *
+	 *	                              +-----------------------+
+	 *	                              | url: 'http://foo.com' |
+	 *	                              +-----------------------+
+	 *	                                      PanelModel
+	 *	                                ^                   +
+	 *	                                |                   |  Input field is
+	 *	                  User clicked  |                   |  in sync with
+	 *	                       "Save".  |                   |  PanelModel#url.
+	 *	                                +                   v
+	 *
+	 *	                            +--------------------------+
+	 *	                            | +----------------------+ |
+	 *	                            | |http://foo.com        | |
+	 *	                            | +----------------------+ |
+	 *	                            |                   +----+ |
+	 *	                            |                   |Save| |
+	 *	                            |                   +----+ |
+	 *	                            +--------------------------+
+	 * @private
+	 * @returns {link.ui.LinkBalloonPanel} Link balloon panel instance.
+	 */
+	_createBalloonPanel() {
+		const editor = this.editor;
+		const viewDocument = editor.editing.view;
+		const linkCommand = editor.commands.get( 'link' );
+
+		// Create the model of the panel.
+		const panelModel = new Model( {
+			maxWidth: 300,
+			url: linkCommand.value
+		} );
+
+		// Bind panel model to command.
+		panelModel.bind( 'url' ).to( linkCommand, 'value' );
+
+		// Create Balloon panel instance.
+		const balloonPanel = new LinkBalloonPanel( panelModel, new LinkBalloonPanelView( editor.locale ) );
+
+		// Observe `LinkBalloonPanelMode#execute` event from within the model of the panel,
+		// which means that the `Save` button has been clicked.
+		this.listenTo( panelModel, 'execute', () => {
+			editor.execute( 'link', balloonPanel.urlInput.value );
+			balloonPanel.view.hide();
+		} );
+
+		// Observe `LinkBalloonPanelMode#execute-unlink` event from within the model of the panel,
+		// which means that the `Unlink` button has been clicked.
+		this.listenTo( panelModel, 'execute-unlink', () => {
+			editor.execute( 'unlink' );
+			balloonPanel.view.hide();
+		} );
+
+		// Always focus editor on panel hide.
+		this.listenTo( balloonPanel.view.model, 'change:isVisible', ( evt, propertyName, value ) => {
```

Comment:
It is kind of workaround for lack of something as focus manager. BalloonPanel still needs some enhancements like close after clicking outside of it. I propose to create a follow up for introducing focus manager and improving this behaviour.